### PR TITLE
[REVIEW] Add `max_file_size` parameter to chunked parquet dataset writer

### PR DESCRIPTION
**Pull request by galipremsagar**
Resolves: #10144 

This PR introduces a parameter, `max_file_size` which when specified the parquet dataset writer will create multiple parquet files that are always less than `max_file_size`.

### CODE REVIEW

File: python/cudf/cudf/io/parquet.py [hunk 1]
```diff
@@ -16,6 +16,26 @@
...
+    "TiB": 2**40,
+    "PiB": 2**50,
+    "B": 1,
+    "": 1,
+}
+BYTE_SIZES = {k.lower(): v for k, v in BYTE_SIZES.items()}
+BYTE_SIZES.update(
+    {k[0]: v for k, v in BYTE_SIZES.items() if k and "i" not in k}
+)
+BYTE_SIZES.update({k[:-1]: v for k, v in BYTE_SIZES.items() if k and "i" in k})
```
Comment (brandon-b-miller) [inline python/cudf/cudf/io/parquet.py:37]: If we're going to start with a hardcoded dict why not just hardcode the result of this processing?

```
{'': 1,
 'b': 1,
 'g': 1000000000,
 'gb': 1000000000,
 'gi': 1073741824,
 'gib': 1073741824,
 'k': 1000,
 'kb': 1000,
 'ki': 1024,
 'kib': 1024,
 'm': 1000000,
 'mb': 1000000,
 'mi': 1048576,
 'mib': 1048576,
 'p': 1000000000000000,
 'pb': 1000000000000000,
 'pi': 1125899906842624,
 'pib': 1125899906842624,
 't': 1000000000000,
 'tb': 1000000000000,
 'ti': 1099511627776,
 'tib': 1099511627776}
```

File: python/cudf/cudf/io/parquet.py [hunk 2]
```diff
@@ -712,9 +738,92 @@ def _get_partitioned(
     return full_paths, metadata_file_paths, grouped_df, part_offsets, filename
 
 
+@_cudf_nvtx_annotate
+def _get_partitioned_new(
```
Comment (brandon-b-miller) [inline python/cudf/cudf/io/parquet.py:742]: is this function supposed to supersede `_get_partitioned`? Can the other one be removed?

File: python/cudf/cudf/io/parquet.py [hunk 3]
```diff
@@ -712,9 +738,92 @@ def _get_partitioned(
...
+    # Copy the entire keys df in one operation rather than using iloc
+    part_names = part_names.to_pandas().to_frame(index=False)
+
+    return part_names, grouped_df, part_offsets
+
+
 ParquetWriter = libparquet.ParquetWriter
 
 
+def _parse_bytes(s):
```
Comment (brandon-b-miller) [inline python/cudf/cudf/io/parquet.py:767]: should we note the file this was vendored from dask in case we need to update it?

File: python/cudf/cudf/io/parquet.py [hunk 4]
```diff
@@ -791,27 +911,110 @@ def __init__(
...
+                        )
+                        > self.max_file_size
+                    ):
+                        curr_file_num += 1
+                        new_file_name = (
+                            self.filename
+                            + "_"
+                            + str(curr_file_num)
+                            + ".parquet"
+                        )
```
Comment (brandon-b-miller) [inline python/cudf/cudf/io/parquet.py:990]: I think it'd be a little clearer like

```
new_file_name = f"{self.filename}_{curr_file_num}.parquet"
```

File: python/cudf/cudf/io/parquet.py [hunk 5]
```diff
@@ -791,27 +911,110 @@ def __init__(
...
+                    new_offsets.append(end)
+                    num_chunks = len(new_offsets)
+                    full_offsets.extend(new_offsets)
+                else:
+                    full_offsets.append(end)
 
+                curr_file_num = 0
+                while num_chunks > 0:
+                    new_file_name = (
+                        self.filename + "_" + str(curr_file_num) + ".parquet"
```
Comment (brandon-b-miller) [inline python/cudf/cudf/io/parquet.py:973]: Same comment as below using an fstring

**Comment by brandon-b-miller** [comment]
Just did a cursory review so far, this is really cool! playing with it a bit locally.

File: python/cudf/cudf/io/parquet.py (hunk 4 above)
Comment (galipremsagar) [reply python/cudf/cudf/io/parquet.py:985]: Yup, changed it

File: python/cudf/cudf/io/parquet.py (hunk 3 above)
Comment (galipremsagar) [reply python/cudf/cudf/io/parquet.py:765]: Yup, mentioned it in docstring

File: python/cudf/cudf/io/parquet.py (hunk 2 above)
Comment (galipremsagar) [reply python/cudf/cudf/io/parquet.py]: Nope, it is just extracting the necessary logic that can be shared, renamed it now to `_get_groups_and_offsets`

File: python/cudf/cudf/io/parquet.py (hunk 1 above)
Comment (galipremsagar) [reply python/cudf/cudf/io/parquet.py]: Changed this to a pre-populated dict.

**Comment by galipremsagar** [comment]
> Just did a cursory review so far, this is really cool! playing with it a bit locally.

Thanks @brandon-b-miller ! This should now be ready for another round of review.

File: python/cudf/cudf/io/parquet.py [hunk 6]
```diff
@@ -672,6 +697,12 @@ def _generate_filename():
     return uuid4().hex + ".parquet"
 
 
+def get_estimated_file_size(df):
+    df_mem_usage = df.memory_usage().sum()
+    file_size = df_mem_usage // 2.5
```
Comment (brandon-b-miller) [inline python/cudf/cudf/io/parquet.py:702]: was just curious where does this heuristic come from? Is it based off a known compression ratio?

File: python/cudf/cudf/io/parquet.py [hunk 7]
```diff
@@ -791,27 +911,105 @@ def __init__(
...
+                    f"{name}={val}"
+                    for name, val in zip(self.partition_cols, keys)
+                ]
+            )
+            prefix = fs.sep.join([self.path, subdir])
+            fs.mkdirs(prefix, exist_ok=True)
+            current_offset = (part_offsets[idx], part_offsets[idx + 1])
+            num_chunks = 1
+            if self.max_file_size is not None:
+                start, end = current_offset
```
Comment (brandon-b-miller) [inline python/cudf/cudf/io/parquet.py:955]:
```suggestion
                # get the current partition
                start, end = current_offset
```

File: python/cudf/cudf/io/parquet.py [hunk 8]
```diff
@@ -791,27 +911,105 @@ def __init__(
...
+                sliced_df = grouped_df[start:end]
+
+                current_file_size = get_estimated_file_size(sliced_df)
+                if current_file_size > self.max_file_size:
+                    parts = int(current_file_size // self.max_file_size)
```
Comment (brandon-b-miller) [inline python/cudf/cudf/io/parquet.py:960]:
```suggestion
                    # if the file is too large, compute metadata for smaller chunks
                    parts = int(current_file_size // self.max_file_size)
```

File: python/cudf/cudf/io/parquet.py [hunk 9]
```diff
@@ -791,27 +911,105 @@ def __init__(
...
+                    new_file_name = f"{self.filename}_{curr_file_num}.parquet"
+                    new_full_path = fs.sep.join([prefix, new_file_name])
+                    while (
+                        new_full_path in self._file_sizes
+                        and (
+                            self._file_sizes[new_full_path]
+                            + (current_file_size / num_chunks)
+                        )
+                        > self.max_file_size
+                    ):
```
Comment (brandon-b-miller) [inline python/cudf/cudf/io/parquet.py:981]: This is the only section I found a little hard to follow. I understand we need a new filepath for each chunk, but I'm having trouble seeing the conditions under which we'd already find the filepath inside `self._file_sizes`. Could you help me out @galipremsagar ?

File: python/cudf/cudf/tests/test_parquet.py [hunk 10]
```diff
@@ -1696,6 +1696,53 @@ def test_parquet_writer_chunked_partitioned(tmpdir_factory, return_meta):
     assert_eq(got_pd, got_cudf)
 
 
+@pytest.mark.parametrize("max_file_size", [100, "10KB"])
+def test_parquet_writer_chunked_max_file_size(tmpdir_factory, max_file_size):
```
Comment (brandon-b-miller) [inline python/cudf/cudf/tests/test_parquet.py:1700]: my only question about this is that it seems like this checks that we roundtrip the data in the case where `max_file_size` is passed, but doesn't as far as I can see do any inspection of the file sizes on disk. is that something that is worth checking or could be done simply?

**Comment by brandon-b-miller** [comment]
Left a few more comments. Overall really easy to follow, I made some suggestions with comments that helped me keep track of things a bit, feel free to reject these if you think it's clear enough already. Just had trouble with one small section.

File: python/cudf/cudf/io/parquet.py (hunk 6 above)
Comment (galipremsagar) [reply python/cudf/cudf/io/parquet.py:702]: So, this was based on experiments on a data frame with no nulls and all unique values for almost all datatypes, the size of file will translate to 2.5x times the GPU memory usage once loaded into the GPU. Obviously, this is not a final heuristic that will work for all types of datasets, but we can improve it in the future or give an undocumented parameter so that customers can tweak it according to their needs if they see it under-cut a lot.

File: python/cudf/cudf/io/parquet.py (hunk 9 above)
Comment (galipremsagar) [reply python/cudf/cudf/io/parquet.py:983]: So when we call `write_table` from the same instance on two different data frames, it could be possible that a file name is already taken by a part of the first dataframe, hence we need this check for the second/future data frame to pick a new file name that doesn't exist in `self._file_sizes`

Comment (galipremsagar) [reply python/cudf/cudf/io/parquet.py:987]: Added the same in the comment

File: python/cudf/cudf/tests/test_parquet.py (hunk 10 above)
Comment (galipremsagar) [reply python/cudf/cudf/tests/test_parquet.py]: Yup, added validation. I bumped the file sizes a bit since the current file sizes are too small for parquet file creation.

**Review by brandon-b-miller — APPROVED** [review]
This LGTM.

File: python/cudf/cudf/tests/test_parquet.py [hunk 11]
```diff
@@ -1696,6 +1697,65 @@ def test_parquet_writer_chunked_partitioned(tmpdir_factory, return_meta):
...
+        got_pd.sort_values(["b"]).reset_index(drop=True),
+        got_cudf.sort_values(["b"]).reset_index(drop=True),
+    )
+
+    all_files = glob.glob(gdf_dir + "/**/*.parquet", recursive=True)
+    for each_file in all_files:
+        # Validate file sizes with some extra 1000
+        # bytes buffer to spare
+        assert os.path.getsize(each_file) <= (
+            max_file_size_in_bytes + 1000
```
Comment (vuule) [inline python/cudf/cudf/tests/test_parquet.py:1747]: can we derive 1000 from the data(frame) somehow?
Otherwise the test is a bit fragile.

File: python/cudf/cudf/io/parquet.py [hunk 12]
```diff
@@ -712,9 +736,94 @@ def _get_partitioned(
...
+    >>> _parse_bytes('5 foos')
+    Traceback (most recent call last):
+        ...
+    ValueError: Could not interpret 'foos' as a byte unit
+    """
+    if isinstance(s, (int, float)):
+        return int(s)
+    s = s.replace(" ", "")
+    if not any(char.isdigit() for char in s):
+        s = "1" + s
```
Comment (vuule) [inline python/cudf/cudf/io/parquet.py:799]: can you give an example where this is used?

File: python/cudf/cudf/io/parquet.py [hunk 13]
```diff
@@ -791,27 +911,111 @@ def __init__(
...
+            else:
+                self.filename = self.filename or _generate_filename()
+                full_path = fs.sep.join([prefix, self.filename])
+                full_paths.append(full_path)
+                metadata_file_paths.append(
+                    fs.sep.join([subdir, self.filename])
+                )
+                full_offsets.append(current_offset[1])
 
+        # return full_paths, metadata_file_paths, grouped_df,
```
Comment (vuule) [inline python/cudf/cudf/io/parquet.py:1012]: is return the right word here?

File: python/cudf/cudf/io/parquet.py [hunk 14]
```diff
@@ -791,27 +911,111 @@ def __init__(
...
+            if self.max_file_size is not None:
+                # get the current partition
+                start, end = current_offset
+                sliced_df = grouped_df[start:end]
+
+                current_file_size = get_estimated_file_size(sliced_df)
+                if current_file_size > self.max_file_size:
+                    # if the file is too large, compute metadata for
+                    # smaller chunks
+                    parts = int(current_file_size // self.max_file_size)
```
Comment (vuule) [inline python/cudf/cudf/io/parquet.py:963]: should we round this up?
If the current file size is 1.5x of the max file size, we need two parts, not one.

**Review by vuule — COMMENTED** [review]
Cool feature.
A few minor comments.

File: python/cudf/cudf/io/parquet.py (hunk 12 above)
Comment (galipremsagar) [reply python/cudf/cudf/io/parquet.py:799]: In this case, `_parse_bytes('MB')`

File: python/cudf/cudf/io/parquet.py (hunk 14 above)
Comment (galipremsagar) [reply python/cudf/cudf/io/parquet.py]: Fixed this 👍

File: python/cudf/cudf/io/parquet.py (hunk 13 above)
Comment (galipremsagar) [reply python/cudf/cudf/io/parquet.py]: Dead code, removed.

File: python/cudf/cudf/tests/test_parquet.py (hunk 11 above)
Comment (galipremsagar) [reply python/cudf/cudf/tests/test_parquet.py]: Yeah, I fixed some issues in the guesstimation function and tweaked the test range a bit so we check realistic sizes.

File: python/cudf/cudf/io/parquet.py [hunk 15]
```diff
@@ -973,7 +974,9 @@ def write_table(self, df):
                 if current_file_size > self.max_file_size:
                     # if the file is too large, compute metadata for
                     # smaller chunks
-                    parts = int((current_file_size / self.max_file_size) + 0.5)
+                    parts = int(
+                        math.ceil(current_file_size / self.max_file_size)
+                    )
```
Comment (vuule) [inline python/cudf/cudf/io/parquet.py:979]: is `int` still needed here?

File: python/cudf/cudf/tests/test_parquet.py [hunk 16]
```diff
@@ -1696,6 +1697,66 @@ def test_parquet_writer_chunked_partitioned(tmpdir_factory, return_meta):
     assert_eq(got_pd, got_cudf)
 
 
+@pytest.mark.parametrize(
+    "max_file_size,max_file_size_in_bytes",
+    [("500KB", 500000), ("1000KB", 1000000)],
```
Comment (vuule) [inline python/cudf/cudf/tests/test_parquet.py:1702]: Just to cover the size parsing code
```suggestion
    [("500KB", 500000), ("MB", 1000000)],
```

**Review by vuule — APPROVED** [review]
Looks good. Left a few optional nitpicks.

File: python/cudf/cudf/io/parquet.py (hunk 15 above)
Comment (galipremsagar) [reply python/cudf/cudf/io/parquet.py]: Right, this is redundant. Removed it.

**Comment by galipremsagar** [comment]
@gpucibot merge
